docs: clarify generic vs charmed alert rules (#272)
canonical/observability-stack: docs/explanation/alerting/alerting-overview.md

---
myst:
 html_meta:

>   description: "Overview of alert rules, their objectives, and ."

Reviewer: don't forget to add this :)

>   description: "Overview of alert rules, their objectives, and ."

Reviewer 2: and..?? What a cliff hanger!

---

# Alert rules

> ## Overview

Reviewer: I'd add an intro between the title and Overview - a quick intro to what alert roles are / how COS uses them (then the reader can decide if they want to continue / learn more)

something like:
"Alert rules define the conditions under which COS raises alerts based on
metrics or logs collected from workloads. When a condition is met for a
specified period, the rule produces an alert, which is then routed by
Alertmanager to the configured receivers."

(how I wrote it pulls some information out of your overview, so that section might change slightly as well)

Alert rules define the conditions under which Prometheus, Mimir, or Loki will fire an alert to Alertmanager. Each rule contains an expression that is evaluated on a schedule; when the expression is true for long enough (the `for` duration), the alert transitions from `pending` to `firing` and Alertmanager routes a notification to the configured channels.

> Alert rules are stored in YAML files and grouped by name. Each rule carries metadata such as a name, severity label, and human-readable annotations that Alertmanager and Grafana use to present actionable notifications. In COS, the [Juju topology](/explanation/architecture/juju-topology.md) is used to annotate an alert rule with identifiers such as `juju_application`, `juju_model`, `juju_model_uuid`, `juju_unit`, and `juju_charm`.

Reviewer 2: Reorg "from zero to hero"

Author: I've done this reorg based on suggestions by Yanisa. In short, before the `overview` section, I've added a brief "zero" intro to alert rules, and continued it inside `overview`.

> As of writing, there is no official centralised store for alert rules; the [Awesome Prometheus Rules](https://samber.github.io/awesome-prometheus-alerts/) project is a commonly used starting point for generic rules.

Reviewer 2: Rephrase: each team does its own alert rules because it's difficult to have one size fits all, but with charms and COS, ...

Author: WDYT about:
"""
There is no official centralized store for alert rules; the [Awesome Prometheus Rules](https://samber.github.io/awesome-prometheus-alerts/) project is a commonly used starting point for generic rules. Because alerting thresholds and priorities vary across teams and workloads, a one-size-fits-all ruleset is not used, however, with COS and charmed operators, teams can ship their own alert rules alongside their workloads and have them automatically loaded into COS without manual configuration.
"""

> As of writing, there is no official centralised store for alert rules; the [Awesome Prometheus Rules](https://samber.github.io/awesome-prometheus-alerts/) project is a commonly used starting point for generic rules.

Reviewer: ```suggestion
There is no official centralized store for alert rules; the [Awesome Prometheus Rules](https://samber.github.io/awesome-prometheus-alerts/) project is a commonly used starting point for generic rules.
```

> Currently, COS supports alerts based on metrics and logs, which are commonly referred to as PromQL and LogQL rules respectively.

Reviewer: You should also add something brief about charmed vs generic alert rules. Something like "COS supports three different sources for alert rules: generic, charmed, and COS config, and alerts based on metrics or logs."

Author: I've added this (well, roughly this :)) in the `Sources of alert rules` section. Do you think it needs an explicit mention or statement?

> ### PromQL alert rules

Reviewer: See in my example headings below - I suggested to generalize it more (Metrics-based alert rules (PromQL))

If they're really just referred to as "PromQL" though, then keep the section titles how you have them (specifying not X-based - just PromQL/LogQL)

Author: Let's go with your previous suggestion i.e. `Metrics-based PromQL`. In conversation, we do usually refer to them as PromQL rules rather than `metrics-based rules`, but I think how you had suggested it before is more explicit and clear in docs.

PromQL (Prometheus Query Language) is used to write alert rules that operate on metrics. These rules are evaluated by Prometheus and Mimir. An expression selects one or more time series and applies filters, aggregations, or arithmetic to produce a scalar result. When the result satisfies the condition (e.g. `> 0.5`), the alert fires.

> A typical PromQL alert rule looks like this:

Reviewer: ```suggestion
A typical PromQL alert rule looks like:
```

> ```yaml

Reviewer 2: I wonder if we should link out for the basic intro, and give some juju delta from our perspective here.

Author: I've added this at the bottom of the PromQL section:
"""
For more information on PromQL, refer to the official Prometheus docs on [querying](https://prometheus.io/docs/prometheus/latest/querying/basics/).
"""
I'll add a corresponding one for LogQL, but I'm leaning towards keeping the example of the rule along with its explanation here. I can also remove it if you think it's going well beyond what should be within docs for COS, but I think this explanation can be helpful.
WDYT @YanisaHS?

Reviewer 2: Agreed.
Should we note here again that they don't need to manually specify topology labels?

alert: HighErrorRate
expr: |
  rate(http_requests_total{status=~"5.."}[5m])
    /
  rate(http_requests_total[5m])
    > 0.05
for: 10m
labels:
  severity: critical
annotations:
  summary: "High HTTP 5xx error rate on {{ $labels.job }}"
  description: "More than 5% of requests are returning 5xx for the past 10 minutes."
```

Here, `rate(...)` computes the per-second rate of requests over a 5-minute window. Dividing error requests by all requests gives a ratio; the alert fires only when that ratio exceeds 5% for 10 consecutive minutes.

> ### LogQL alert rules

Reviewer: Same here - suggest to make it more generic in the heading (Log-based alert rules (LogQL))

LogQL is Loki's query language for log streams. This rules are evaluated by Loki. Alert rules using LogQL are called *log alerts* and rely on metric queries. Log lines are counted or sampled over a time window to produce a numeric value that can be compared against a threshold.

A typical LogQL alert rule looks like this:

```yaml
alert: HighLogErrorRate
expr: |
  sum(rate({job="my-service"} |= "error" [5m])) > 10
for: 5m
labels:
  severity: warning
annotations:
  summary: "Elevated error log rate for {{ $labels.job }}"
  description: "More than 10 error log lines per second have been observed for 5 minutes."
```

Here, `rate(...)` counts log lines matching the filter `|= "error"` per second over a 5-minute window, then `sum(...)` aggregates across all streams for the job. The alert fires when that rate exceeds 10 for 5 continuous minutes.

> ## References

Reviewer: I put this in another comment, but you could move these links within the "Alerts in COS" section and get rid of the references. Since it's just these two links anyway

Author: Agreed, I've removed this section altogether.

- [Charmed alert rules](./charmed-rules): How alert rules bundled with charms are transformed and forwarded within COS.
- [Generic alert rule groups](./generic-rules): Built-in host-health rule groups provided by COS.

> ## Alerts in COS

Author: @sed-i I've left the info under the `Alerts in COS` heading 2 as bullet points. I can alternatively turn each bullet point into an H3? WDYT?

Reviewer 2: I'm leaning towards H3.
Also, should we move this section closer to the top?

Author: I've changed the BP's to H3s and also moved this section to the top, right after overview.

> ## Alerts in COS

Reviewer: IMO I'd keep this as a H2, not make the others H3, but move this up before the PromQL and LogQL parts. I'm suggesting this because to me, the charmed vs. generic items feel less specific than PromQL / LogQL parts. The charmed vs generic part is still more conceptual, whereas PromQL / LogQL is getting closer to implementation (presumably a user would hit the charm/generic concept before the type of metric)

So scaffolding down from big picture info to more specific stuff, I would organize it like:
```
# Alert rules
<short intro - what are alert rules and how does it relate to COS>
## Overview
<overview / high-level description of alert rules>
## Sources of alert rules
<generic, charmed, COS config charm>
## Metrics-based alert rules (PromQL)
<description>
## Logs-based alert rules (LogQL)
<description>
```
OR if you want the H3s
```
# Alert rules
<short intro - what are alert rules and how does it relate to COS>
## Overview
<overview / high-level description of alert rules>
## Sources of alert rules
### Generic alert rules
<description>
### Charmed alert rules
<description>
### COS config charm
<description>
## Type of alert rule
### Metrics-based (PromQL)
<description>
### Logs-based (LogQL)
<description>
```

But regardless, if you reorganize it this way or not, I'd cut the References section and put those links in the Sources of alert rules section

Author: I've gone with your second option so that we use H3s like @sed-i also suggested. I've also removed the References section.

Each rule carries metadata such as a name, severity label, and human-readable annotations that Alertmanager and Grafana use to present actionable notifications. In COS, the [Juju topology](/explanation/architecture/juju-topology.md) is used to annotate alert rules with identifiers such as `juju_application`, `juju_model`, `juju_model_uuid`, `juju_unit`, and `juju_charm`. COS supports alerts based on metrics and logs, commonly referred to as PromQL and LogQL rules respectively.

There is no official centralized store for alert rules; the [Awesome Prometheus Rules](https://samber.github.io/awesome-prometheus-alerts/) project is a commonly used starting point for generic rules. Because alerting thresholds and priorities vary across teams and workloads, a one-size-fits-all ruleset is not used, however, with COS and charmed operators, teams can ship their own alert rules alongside their workloads and have them automatically loaded into COS without manual configuration.

## Sources of alert rules

> COS alert rules come from three sources:

Reviewer 2: ```suggestion
COS alert rules are sourced in several ways.
```

### Generic alert rules

Generic rules are a minimal set of host-health rules shipped with COS itself, covering common failure scenarios such as unreachable targets and missing metrics. They require no configuration and apply automatically across deployments. The main focus of these alerts is the health of COS itself, as well as the reachability and `up` status of monitored instances. See [Generic alert rule groups](./generic-rules) for the full list and their behaviour.

### Charmed alert rules

Charmed rules are bundled directly with a charmed operator, encoding the operational knowledge its author considers important. They are automatically forwarded to Prometheus, Mimir, or Loki over Juju relations when the charm is deployed. They are written by the authors of charms monitored by COS and, unlike generic rules, they focus on specific workload-related health metrics. See [Charmed alert rules](./charmed-rules) for how they are transformed and managed.

> ### COS config charm

Reviewer 2: ```suggestion
### Git-ops alert rules
```

Custom alert rules can be loaded into COS from an external Git repository using the [COS Configuration charm](https://charmhub.io/cos-configuration-k8s). This is the recommended approach for organisation-specific or workload-specific rules that are not bundled with a charm. See [How to sync alert rules from a Git repository](../../how-to/configure-and-tune/sync-alert-rules-from-git) for setup instructions.

> 

Reviewer 2: We should probably mention here the `forward_alert_rules` config option for disabling alert rules propagation.
